src/Spectre.libWavelet/WaveletDecomposerRef.cpp

```diff
@@ -0,0 +1,126 @@
+            coefficients[row][column] = coefficients[i][j];
+        }
+        coefficients[i].resize(oldRowColumn); // Used to fill moved entries with zeros.
+        coefficients[i].resize(blockLength);
+    }
+}
+
+// Applies the high and low pass filters on the current low-frequency
+// wavelet coefficients, at the level determined by the caller.
+// https://upload.wikimedia.org/wikipedia/commons/1/16/Wavelets_-_SWT_Filter_Bank.png
+// The figure above depicts the process, where h_j prepresents a high- and g_j a low-pass
+// filters, applied at level j.
+// All the low frequency coefficients are overridden on each call, while the high
+// frequnecy ones are just added.
+inline void WaveletDecomposerRef::ApplyFilters(WaveletCoefficients& coefficients,
+    CoefficientsPerLevel& lowFrequencyCoefficients, size_t scale, size_t blockLength, unsigned level) const
+{
+    coefficients.data[level].resize(scale); // Adjust number of coefficient lists for that certain level
+    for (unsigned i = 0; i < scale; i++)
+    {
+        lowFrequencyCoefficients[i].resize(blockLength); // Extend with zeros/shrink to match size
+        coefficients.data[level][i] = // Apply Daubechies4 decomposition high pass filter
+            m_Convolution.Convolve(precomputed::DecompositionHighPassFilter, lowFrequencyCoefficients[i]);
+        lowFrequencyCoefficients[i] = // Apply Daubechies4 decomposition low pass filter
+            m_Convolution.Convolve(precomputed::DecompositionLowPassFilter, lowFrequencyCoefficients[i]);
+    }
+}
+
+WaveletDecomposerRef::WaveletDecomposerRef() : m_Convolution()
+{
+};
+
+WaveletCoefficients WaveletDecomposerRef::Decompose(Signal&& signal) const
+{
+    const size_t signalLength = signal.size();
+
+    WaveletCoefficients coefficients;
+    CoefficientsPerLevel& lowFrequencyCoefficients = coefficients.data[WAVELET_LEVELS];
+
+    // The number of samples N used to generate N coefficient lists
+    // for the lowest frequency decomposition.
+    constexpr size_t LOW_FREQ_SAMPLES_NUM = 1 << (WAVELET_LEVELS - 1); // 2^(WL - 1)
+    lowFrequencyCoefficients.resize(LOW_FREQ_SAMPLES_NUM);
+    lowFrequencyCoefficients[0] = std::move(signal); // Initialize first line with signal
+
+    size_t scale = 1; // Initial wavelet scale (or it's frequency)
+    size_t blockLength = ComputeBlockLength(0, scale, signalLength); // Initial number of coefficients
+
+    for (unsigned currentLevel = 0; currentLevel < WAVELET_LEVELS - 1; currentLevel++)
+    {
+        ApplyFilters(coefficients, lowFrequencyCoefficients, scale, blockLength, currentLevel);
+
+        scale = ComputeScale(currentLevel);
+        blockLength = ComputeBlockLength(currentLevel + 1, scale, signalLength);
+        DownsampleByDissolving(lowFrequencyCoefficients, currentLevel + 1, (unsigned)blockLength);
+    }
+
+    ApplyFilters(coefficients, lowFrequencyCoefficients, scale, blockLength, WAVELET_LEVELS - 1);
+
+    return coefficients;
+}
+}
```

Comment:
This file consists mostly of `static inline` functions, that are inivisble anywhere else outside of this file. I know, it's used nowhere else, but limits our capabilities of testing. Let's leave it this way for now, but I suppose you'll need to make this more testable (and to implement tests) before you'll start speeding this up. Just for checking consistence other way than reading values of variables in *Debug*.

Reply:
Yeah well, I do believe that the standard way of hiding the implementation details from the public interface is creating 2 classes, for example:
Denoiser
DenoiserImpl

Where Denoiser has a single public method Denoise which calls the DenoiserImpl method(s).
...And DenoiserImpl has all those actual methods implemented as public for the sake of testing.

The reason I have made them inlines/static was done for the sake of achieving a bit of a speed up during compile and run times.